### PR TITLE
External magnetic field

### PR DESCRIPTION
**Pull request by SchoeniPhlippsn**
## Description

Adds a new constant magnetic field that acts on magnetic dipoles.

## Motivation and context

Allows for a coupling between magnetic particles and an external magnetic field.

## How has this been tested?

I added pytests and tested it on different systems both using CPU and GPU.

## Change log

```
Addition of constant external magnetic field that acts on magnetic dipoles
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/md/EvaluatorExternalMagneticField.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            d["B"] = pybind11::make_tuple(B.x, B.y, B.z);;
+            d["mu"] = pybind11::make_tuple(mu.x, mu.y, mu.z);;
+            return d;
+            }
+#endif // ifndef __HIPCC__
+        } __attribute__((aligned(16)));
+
+    typedef void* field_type;
+
+    //! Constructs the constraint evaluator
```
Comment (tcmoore3) [inline hoomd/md/EvaluatorExternalMagneticField.h]:
```suggestion
    //! Constructs the external field evaluator
```

File: hoomd/md/EvaluatorExternalMagneticField.h [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    /*! No contribution
+     */
+    DEVICE static bool requestFieldVirialTerm()
+        {
+        return false;
+        }
+
+    //! Evaluate the force, energy and virial
+    /*! \param F force vector
+        \param energy value of the energy
```
Comment (tcmoore3) [inline hoomd/md/EvaluatorExternalMagneticField.h]:
```suggestion
    /*! \param F force vector
        \param T torque vector
        \param energy value of the energy
```

File: hoomd/md/external/field.py [hunk 3]
```diff
@@ -87,7 +87,7 @@ class Electric(Field):
     """Electric field force.
 
     `Electric` computes forces, and virials, and energies on all particles in
```
Comment (tcmoore3) [inline hoomd/md/external/field.py]:
```suggestion
    `Electric` computes forces, virials, and energies on all particles in
```

File: hoomd/md/external/field.py [hunk 4]
```diff
@@ -120,3 +120,46 @@ def __init__(self):
...
+
+    `Magnetic` computes torces and energies on all particles in
+    the simulation state which are consistent with: 
+
+    .. math::
+
+       U_i = -\\vec{mu}_i \\cdot \\vec{B}
+
+
+    where :math:`\\vec{mu}_i` is the particle magnetic momentum and
```
Comment (tcmoore3) [inline hoomd/md/external/field.py]:
```suggestion
    where :math:`\\vec{mu}_i` is the magnetic dipole moment of particle :math:`i`
```

File: hoomd/md/external/field.py [hunk 5]
```diff
@@ -120,3 +120,46 @@ def __init__(self):
...
+    .. py:attribute:: params
+
+        The `Magnetic` external potential parameters. The dictionary has the
+        following keys:
+
+        * ``B`` (`tuple` [`float`, `float`, `float`] ,**required**) - The magnetic 
+        field vector :math:`[\\mathrm{energy} \\cdot \\mathrm{time} \\cdot \
+        \\mathrm{charge}^{-1} \\cdot \\mathrm{length}^{-2} ]`.
+        * ``mu`` (`tuple` [`float`, `float`, `float`] ,**required**) - The magnetic
+        moment of the particles type :math:`[\\mathrm{charge} \\cdot \
```
Comment (tcmoore3) [inline hoomd/md/external/field.py]:
```suggestion
        * ``mu`` (`tuple` [`float`, `float`, `float`] ,**required**) - The magnetic
        dipole moment of the particles type :math:`[\\mathrm{charge} \\cdot \
```

File: hoomd/md/external/field.py [hunk 6]
```diff
@@ -120,3 +120,46 @@ def __init__(self):
...
+        \\mathrm{length}^2 \\cdot \\mathrm{time}^{-1}]`.
+
+        Type: `TypeParameter` [``particle_type``, `dict`]
+
```
Comment (tcmoore3) [inline hoomd/md/external/field.py:164]: Add a note that you define `B` in the global reference frame but `mu` in the particle reference frame.

**Comment by tcmoore3** [comment]
pre-commit.ci autofix

**Review by tcmoore3 — APPROVED** [review]
Nice work. I was able to build and test the branch and it seems to work as expected.

It looks like you need to add the new python class to the doc pages so that it shows in the sphinx documentation. There are also some flake8 errors to fix. Otherwise this should be good to go.

**Comment by SchoeniPhlippsn** [comment]
pre-commit.ci autofix

**Comment by SchoeniPhlippsn** [comment]
pre-commit.ci autofix

**Comment by joaander** [comment]
I fixed the readthedocs webhook and docs are now building again. This is failing with:
```
/home/docs/checkouts/readthedocs.org/user_builds/hoomd-blue/checkouts/1637/hoomd/md/external/field.py:docstring of hoomd.md.external.field.Magnetic:21: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

**Review by joaander — APPROVED** [review]
Thanks!

I applied some changes directly to the branch that fixed code style, added tests for the documentation code snippets, and made the documentation format correctly.

**Pull request merged**